crates/btcio/src/reader/query.rs

```diff
@@ -279,7 +299,116 @@ async fn poll_for_new_blocks<R: Reader>(
+            max_wait_time: Duration::from_secs(300), // 5 minutes
+        }
+    }
+}
+
+/// Watches the status channel for chainstate which has the finalized state matching as referred by
+/// checkpoint. If it is found, returns `true`, otherwise returns `false`. `false` is returned when
+/// there is a different chainstate than indicated by checkpoint for given slot.
+async fn wait_until_checkpoint_accepted_or_rejected<R: Reader>(
+    ctx: &ReaderContext<R>,
+    checkpoint: &SignedCheckpoint,
+) -> anyhow::Result<bool> {
+    let l2_commt = checkpoint.checkpoint().batch_info().l2_range.1;
+    let config = &ctx.checkpoint_config;
+
+    info!("Waiting for checkpoint to be accepted");
+    let mut rx = ctx.status_channel.subscribe_chain_sync();
+
+    let start_time = time::Instant::now();
+
+    loop {
+        if start_time.elapsed() > config.max_wait_time {
+            error!("Maximum wait time exceeded waiting for checkpoint acceptance");
+            return Err(anyhow!(
+                "reader:query:wait_for_checkpoint_accept: Maximum wait time of {:?}s exceeded",
+                config.max_wait_time
+            ));
+        } else if let Some(chstate_update) = (*rx.borrow_and_update()).clone() {
+            let tip_commitment = chstate_update.new_status().tip;
+            let same_slot = l2_commt.slot() == tip_commitment.slot();
+            let same_blkid = l2_commt.blkid() == tip_commitment.blkid();
+
+            if same_slot && same_blkid {
+                return Ok(true);
+            } else if same_slot && !same_blkid {
+                warn!("Checkpoint not accepted: different block id for same slot finalized");
+                return Ok(false);
+            } else if l2_commt.slot() < tip_commitment.slot() {
+                // We've been looking for stale checkpoint to be accepted.
+                warn!("Checkpoint not accepted: higher slot finalized than the slot waited for");
+                return Ok(false);
+            }
+        }
+
+        wait_for_chainstate_change(config, &mut rx).await?;
+    }
+}
+
+async fn wait_for_chainstate_change(
+    config: &CheckpointWaitConfig,
+    rx: &mut watch::Receiver<Option<ChainSyncStatusUpdate>>,
+) -> anyhow::Result<()> {
+    // Wait for chainstate change or timeout
+    tokio::select! {
+        result = rx.changed() => {
+            if let Err(e) = result {
+                error!("Status channel closed unexpectedly: {}", e);
+                return Err(anyhow!("reader:query:wait_for_checkpoint_accept: Status channel closed: {}", e));
+            }
+        }
+        _ = tokio::time::sleep(config.chainstate_update_timeout) => {
+            error!("Timed out waiting for chainstate update after {:?}", config.chainstate_update_timeout);
+            return Err(anyhow!(
+                "reader:query:wait_for_checkpoint_accept: Timed out waiting for chainstate update after {:?}s",
+                config.chainstate_update_timeout
+            ));
```

Comment:
Structured errors instead of `anyhow!`.

Also, the status channel returns an error [if-and-only-if the sender is closed, meaning the upstream task is shutting down](https://docs.rs/tokio/latest/tokio/sync/watch/struct.Receiver.html#method.changed).  It's not an error for us, we should just interpret it as a signal that we should exit too.